weed/shell/command_volume_fsck.go

```diff
@@ -504,81 +497,54 @@ func (c *commandVolumeFsck) oneVolumeFileIdsCheckOneVolume(dataNodeId string, vo
-	}
-
-	if *c.verbose {
-		fmt.Fprintln(c.writer, "delete response Status : ", resp.Status)
-		fmt.Fprintln(c.writer, "delete response Headers : ", resp.Header)
-	}
-}
-
-func (c *commandVolumeFsck) oneVolumeFileIdsSubtractFilerFileIds(dataNodeId string, volumeId uint32, collection string) (inUseCount uint64, orphanFileIds []string, orphanDataSize uint64, err error) {
-
-	db := needle_map.NewMemDb()
-	defer db.Close()
-
-	if err = db.LoadFromIdx(getVolumeFileIdFile(c.tempFolder, dataNodeId, volumeId)); err != nil {
+	if err = volumeFileIdDb.LoadFromIdx(getVolumeFileIdFile(c.tempFolder, dataNodeId, volumeId)); err != nil {
 		err = fmt.Errorf("failed to LoadFromIdx %+v", err)
 		return
 	}
 
 	voluemAddr := pb.NewServerAddressWithGrpcPort(dataNodeId, 0)
-	if err = c.readFilerFileIdFile(volumeId, func(nId types.NeedleId, itemPath util.FullPath) {
+	if err = c.readFilerFileIdFile(volumeId, func(filerNeedleId types.NeedleId, itemPath util.FullPath) {
+		// remove from db needles found in filler
+		needleValue, found := volumeFileIdDb.Get(filerNeedleId)
+		if !found {
+			return
+		}
 		inUseCount++
-		if *c.verifyNeedle {
-			if v, ok := db.Get(nId); ok && v.Size.IsValid() {
-				newSize := types.Size(verifyProbeBlobSize)
-				if v.Size > newSize {
-					v.Size = newSize
-				}
-				if _, err := readSourceNeedleBlob(c.env.option.GrpcDialOption, voluemAddr, volumeId, *v); err != nil {
-					fmt.Fprintf(c.writer, "failed to read file %s NeedleBlob %+v: %+v", itemPath, nId, err)
-					if *c.forcePurging {
-						return
-					}
+		if *c.verifyNeedle && needleValue.Size.IsValid() {
+			if _, err := readNeedleStatus(c.env.option.GrpcDialOption, voluemAddr, volumeId, *needleValue); err != nil {
+				fmt.Fprintf(c.writer, "failed to read %d:%s needle status of file %s: %+v\n", volumeId, filerNeedleId.String(), itemPath, err)
+				if *c.forcePurging {
+					return
 				}
 			}
 		}
-
-		if err = db.Delete(nId); err != nil && *c.verbose {
-			fmt.Fprintf(c.writer, "failed to nm.delete %s(%+v): %+v", itemPath, nId, err)
+		if err = volumeFileIdDb.Delete(filerNeedleId); err != nil && *c.verbose {
+			fmt.Fprintf(c.writer, "failed to nm.delete %s(%d:%s): %+v\n", itemPath, volumeId, filerNeedleId.String(), err)
 		}
 	}); err != nil {
 		err = fmt.Errorf("failed to readFilerFileIdFile %+v", err)
 		return
 	}
 
 	var orphanFileCount uint64
-	if err = db.AscendingVisit(func(n needle_map.NeedleValue) error {
-		if !n.Size.IsValid() {
+	cutoffFrom := uint64(c.collectStartTime.Add(-*c.cutoffTimeAgo).UnixNano())
+	// subtract
+	if err = volumeFileIdDb.AscendingVisit(func(n needle_map.NeedleValue) error {
+		// do not mark with orphan already deleted chunks
+		if n.Size.IsDeleted() {
 			return nil
 		}
+		// do not mark with orphan the chunks during the file uploading process
+		if (*c.cutoffTimeAgo).Seconds() > 0 {
+			if needleMeta, err := readNeedleMeta(c.env.option.GrpcDialOption, voluemAddr, volumeId, n); err == nil && cutoffFrom >= needleMeta.LastModified {
```

Comment:
this will cause one remote read for each needle?

Reply:
> this will cause one remote read for each needle?

Yes, but after subtracting the needles found in the filler.

Reply:
the original approach was to read only needles before the cutoff time. what is the reason to change it?

Reply:
> the original approach was to read only needles before the cutoff time. what is the reason to change it?

the original approach had two drawbacks
1. Remote call was for each needle and for the operation of finding lost files in the filler
2. broken needles were removed when copying

Reply:
1. the original approach read only needles before the cutoff time in one grpc call. It is not for each needle.

Reply:
> 1. the original approach read only needles before the cutoff time in one grpc call. It is not for each needle.

It's a delusion

```
> volume.fsck -v
working directory: /var/folders/zv/tqq_jjhx3sb04kjw8565qmf00000gn/T/sw_fsck96629715
collecting volume id and locations from master ...
dn 192.168.1.85:8080 collected 21 volumes and locations.
collecting volume 3 file ids from 192.168.1.85:8080 ...
collecting volume 5 file ids from 192.168.1.85:8080 ...
collecting volume 6 file ids from 192.168.1.85:8080 ...
collecting volume 1 file ids from 192.168.1.85:8080 ...
collecting volume 19 file ids from 192.168.1.85:8080 ...
collecting volume 17 file ids from 192.168.1.85:8080 ...
collecting volume 2 file ids from 192.168.1.85:8080 ...
collecting volume 4 file ids from 192.168.1.85:8080 ...
collecting volume 21 file ids from 192.168.1.85:8080 ...
collecting volume 20 file ids from 192.168.1.85:8080 ...
Failed to search for last valid index on volume 20 with error to read needle meta with id 195 from volume 20 with error rpc error: code = Unknown desc = index out of range 3collecting volume 7 file ids from 192.168.1.85:8080 ...
collecting volume 15 file ids from 192.168.1.85:8080 ...
collecting volume 16 file ids from 192.168.1.85:8080 ...
collecting volume 18 file ids from 192.168.1.85:8080 ...
dn 192.168.1.85:8080 filtred 14 volumes and locations.
checking each file from filer path /...
checking directory /buckets
checking directory /topics
checking directory /buckets/test
checking directory /buckets/test2
checking directory /topics/.system
checking directory /topics/.system/log

Total           entries:200     orphan:0        0.00%   0B
This could be normal if multiple filers or no filers are used.
no orphan data
> 

```

```
I1019 10:36:40.759332 filer_grpc_server_sub_meta.go:279 + listener filer:192.168.1.85:8888@192.168.1.85:63937
I1019 10:36:40.759378 filer_grpc_server_sub_meta.go:106  + filer:192.168.1.85:8888@192.168.1.85:63937 local subscribe / from 2022-10-19 10:35:39.018704 +0500 +05 clientId:-357791754
I1019 10:36:40.759411 filer_grpc_server_sub_meta.go:119 read on disk filer:192.168.1.85:8888@192.168.1.85:63937 local subscribe / from 2022-10-19 10:35:39.018704 +0500 +05
I1019 10:36:40.759469 filer_grpc_server_sub_meta.go:138 read in memory filer:192.168.1.85:8888@192.168.1.85:63937 local subscribe / from 2022-10-19 10:35:39.018704 +0500 +05
I1019 10:40:20.870335 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2617112091 last_modified:1666099768 crc:2923798200
I1019 10:40:20.871423 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:395026492 last_modified:1666099771 crc:2923798200
I1019 10:40:20.871791 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2707242667 last_modified:1666099771 crc:2923798200
I1019 10:40:20.879692 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1963732394 last_modified:1666099774 crc:2923798200
I1019 10:40:20.918254 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:4211117316 last_modified:1666099764 crc:2923798200
I1019 10:40:20.918882 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3918498941 last_modified:1666099770 crc:2923798200
I1019 10:40:20.919403 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1464142951 last_modified:1666099775 crc:2923798200
I1019 10:40:20.919795 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3634573112 last_modified:1666099775 crc:2923798200
I1019 10:40:20.974772 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:117507735 last_modified:1666099764 crc:2923798200
I1019 10:40:20.975788 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:578029757 last_modified:1666099769 crc:2923798200
I1019 10:40:20.976770 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1949010426 last_modified:1666099772 crc:2923798200
I1019 10:40:20.977481 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1230120990 last_modified:1666099773 crc:2923798200
I1019 10:40:21.025086 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:110609058 last_modified:1666099762 crc:2923798200
I1019 10:40:21.025795 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2194251860 last_modified:1666099769 crc:2923798200
I1019 10:40:21.026481 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3853369932 last_modified:1666099775 crc:2923798200
I1019 10:40:21.026927 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:352570335 last_modified:1666099775 crc:2923798200
I1019 10:40:21.027552 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:440116300 last_modified:1666099776 crc:2923798200
I1019 10:40:21.040706 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3236811919 last_modified:1666099793 crc:2923798200
I1019 10:40:21.041531 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:4253057011 last_modified:1666099800 crc:2923798200
I1019 10:40:21.042033 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2196844287 last_modified:1666099805 crc:2923798200
I1019 10:40:21.042703 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2130287089 last_modified:1666099806 crc:2923798200
I1019 10:40:21.058248 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2944431882 last_modified:1666099798 crc:2923798200
I1019 10:40:21.058787 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1487940564 last_modified:1666099804 crc:2923798200
I1019 10:40:21.059293 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1591054643 last_modified:1666099807 crc:2923798200
I1019 10:40:21.059706 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:309446853 last_modified:1666099809 crc:2923798200
I1019 10:40:21.072684 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3565500862 last_modified:1666099758 crc:2923798200
I1019 10:40:21.073276 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:947264582 last_modified:1666099761 crc:2923798200
I1019 10:40:21.073778 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1454843398 last_modified:1666099766 crc:2923798200
I1019 10:40:21.074152 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:279612538 last_modified:1666099772 crc:2923798200
I1019 10:40:21.121281 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:296343357 last_modified:1666099798 crc:2923798200
I1019 10:40:21.122074 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:162088026 last_modified:1666099806 crc:2923798200
I1019 10:40:21.122823 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3106405145 last_modified:1666099808 crc:2923798200
I1019 10:40:21.123620 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3304136432 last_modified:1666099810 crc:2923798200
I1019 10:40:21.124184 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:668780424 last_modified:1666099811 crc:2923798200
I1019 10:40:21.143556 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:223713575 last_modified:1666099802 crc:2923798200
I1019 10:40:21.144276 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:874109879 last_modified:1666099805 crc:2923798200
I1019 10:40:21.144981 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2420460579 last_modified:1666099808 crc:2923798200
I1019 10:40:21.145718 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2204456324 last_modified:1666099809 crc:2923798200
I1019 10:40:21.170045 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1791490484 last_modified:1666099800 crc:2923798200
I1019 10:40:21.170591 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1249976608 last_modified:1666099810 crc:2923798200
I1019 10:40:21.171093 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:112673293 last_modified:1666099811 crc:2923798200
I1019 10:40:21.171459 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2854294950 last_modified:1666099811 crc:2923798200
I1019 10:40:21.183804 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2546464181 last_modified:1666099797 crc:2923798200
I1019 10:40:21.184453 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:254720911 last_modified:1666099808 crc:2923798200
I1019 10:40:21.185122 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:1516661432 last_modified:1666099809 crc:2923798200
I1019 10:40:21.185574 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3618293872 last_modified:1666099810 crc:2923798200
I1019 10:40:21.201667 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3979834063 last_modified:1666099801 crc:2923798200
I1019 10:40:21.202244 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2513236402 last_modified:1666099806 crc:2923798200
I1019 10:40:21.202828 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:2246225349 last_modified:1666099809 crc:2923798200
I1019 10:40:21.203162 volume_grpc_read_write.go:53 volume grpc call read needle metadata  cookie:3141026729 last_modified:1666099811 crc:2923798200

```

Reply:
cc @lyang24 

It could be a bug. But the algo should be reading only once from the server.

Reply:
> cc @lyang24
> 
> It could be a bug. But the algo should be reading only once from the server.
@chrislusf 
And how it should work then I do not understand?
This is possible if you read the last needle in the hope that the modification time is the creation time and is set once during creation and always, new needles are appended to the end of the file. This is true ?

Reply:
and `last_modified:1666099775` Not like nanoseconds https://github.com/seaweedfs/seaweedfs/blob/master/weed/shell/command_volume_fsck.go#L145
```
Failed to search for last valid index on volume 20 cutoffFrom 1666158980835215000 with error to read needle meta with id 195 from volume 20 with error rpc error: code = Unknown desc = index out of range 3collecting volume 5 file ids from 192.168.1.85:8080 ...
```